### ISSUE TITLE
Five syntax analyzers bind every invocation in the compilation: add name pre-filters before GetSymbolInfo and thread CancellationToken through Common helpers

### ISSUE THREAD

**Issue by rjmurillo** [rjmurillo/moq.analyzers#1259]
## Summary

An internal code audit found that five analyzers register `RegisterSyntaxNodeAction(..., SyntaxKind.InvocationExpression)` and immediately call `SemanticModel.GetSymbolInfo(...)` on **every invocation expression in the compilation** — including the overwhelming majority that cannot possibly be the Moq method the analyzer cares about. Semantic binding is the single most expensive step in these callbacks, and it runs on every keystroke in the IDE. The repository already has a sanctioned pattern for this: `SemanticModelExtensions.IsMoqFluentInvocation` (`src/Common/SemanticModelExtensions.cs:254-290`) rejects by method-name text *before* binding, with the symbol check remaining authoritative.

Additionally (audit item L11), several `GetSymbolInfo`/`GetTypeInfo` call sites in these analyzers and in `src/Common` omit the analysis `CancellationToken`, so binding work cannot be cancelled when the host cancels analysis (e.g., on every keystroke). Sibling code threads the token correctly (e.g., `SemanticModelExtensions.FindSetupMethodFromCallbackInvocation` at `src/Common/SemanticModelExtensions.cs:30`, `EventSyntaxExtensions.ValidateEventArgumentTypes` at `src/Common/EventSyntaxExtensions.cs:37`).

**Zero behavior change is intended.** All existing tests must pass unchanged.

Affected analyzers:

| Analyzer | Rule | Unfiltered bind site |
|---|---|---|
| `ReturnsAsyncShouldBeUsedForAsyncMethodsAnalyzer` | Moq1206 | `src/Analyzers/ReturnsAsyncShouldBeUsedForAsyncMethodsAnalyzer.cs:93` |
| `ReturnsDelegateShouldReturnTaskAnalyzer` | Moq1208 | `src/Analyzers/ReturnsDelegateShouldReturnTaskAnalyzer.cs:93` |
| `SetupShouldNotIncludeAsyncResultAnalyzer` | Moq1201 | `src/Analyzers/SetupShouldNotIncludeAsyncResultAnalyzer.cs:66` |
| `RaiseEventArgumentsShouldMatchEventSignatureAnalyzer` | Moq1202 | `src/Analyzers/RaiseEventArgumentsShouldMatchEventSignatureAnalyzer.cs:84` |
| `EventSetupHandlerShouldMatchEventTypeAnalyzer` | Moq1205 | `src/Analyzers/EventSetupHandlerShouldMatchEventTypeAnalyzer.cs:68` |

Correction to the audit report: `SetupShouldNotIncludeAsyncResultAnalyzer` already threads `context.CancellationToken` into both of its `GetSymbolInfo` calls (`:66`, `:78`); it only needs the name pre-filter. The cancellation-token gap applies to the other four analyzers and to the Common helpers listed below.

### The sanctioned pattern this issue replicates

`src/Common/SemanticModelExtensions.cs:254-290` (`IsMoqFluentInvocation`) — note the doc remark:

```csharp
    /// <remarks>
    /// The <paramref name="fastPathNames"/> string check is an intentional optimization,
    /// not detection logic. The <paramref name="symbolPredicate"/> is the authoritative
    /// gate for correctness. ...
    /// </remarks>
```

and its body rejects on `memberAccess.Name.Identifier.ValueText` before calling `GetSymbolInfo` (`:266-283`).

### Example of the problem (Moq1206)

`src/Analyzers/ReturnsAsyncShouldBeUsedForAsyncMethodsAnalyzer.cs:84-107`:

```csharp
    private static bool IsReturnsMethodCallWithAsyncLambda(InvocationExpressionSyntax invocation, SemanticModel semanticModel, MoqKnownSymbols knownSymbols)
    {
        if (invocation.Expression is not MemberAccessExpressionSyntax)
        {
            return false;
        }

        // Query the invocation (not the MemberAccessExpressionSyntax) so Roslyn has argument context
        // for overload resolution. Fall back to CandidateSymbols for delegate overloads.
        SymbolInfo symbolInfo = semanticModel.GetSymbolInfo(invocation);
```

Every invocation in the compilation whose expression is a member access (i.e., nearly all of them) gets bound, even `Console.WriteLine(...)`.

### Cancellation-token gaps (verified, current line numbers)

Analyzer files:

- `ReturnsAsyncShouldBeUsedForAsyncMethodsAnalyzer.cs`: `GetSymbolInfo` at `:93`, `:131`, `:144` — no token.
- `ReturnsDelegateShouldReturnTaskAnalyzer.cs`: `GetSymbolInfo` at `:93`, `:149`, `:180`, `:236`; `GetTypeInfo` at `:270` — no token.
- `RaiseEventArgumentsShouldMatchEventSignatureAnalyzer.cs`: `GetSymbolInfo` at `:84` — no token.
- `EventSetupHandlerShouldMatchEventTypeAnalyzer.cs`: `GetSymbolInfo` at `:68`, `:131`, `:175`; `GetTypeInfo` at `:184`, `:191` — no token.

Common helpers (audit L11):

- `src/Common/SemanticModelExtensions.cs:160` — `GetAllMatchingSymbols<T>`: `SymbolInfo symbolInfo = semanticModel.GetSymbolInfo(expression);`
- `src/Common/SemanticModelExtensions.cs:229` — `TryGetEventSymbolFromLambdaSelector`: `SymbolInfo symbolInfo = semanticModel.GetSymbolInfo(memberAccess);`
- `src/Common/SemanticModelExtensions.cs:283` — `IsMoqFluentInvocation`: `SymbolInfo symbolInfo = semanticModel.GetSymbolInfo(memberAccess);`
- `src/Common/InvocationExpressionSyntaxExtensions.cs:42` — `FindSetupInvocation`: `SymbolInfo symbolInfo = semanticModel.GetSymbolInfo(candidateInvocation);`
- `src/Common/SyntaxNodeExtensions.cs:22` — `FindLocation<TSyntax>`: `ISymbol? symbol = semanticModel?.GetSymbolInfo(n).Symbol;`

## Desired behavior

1. Each of the five analyzers rejects invocations by member-name text before any semantic binding. The name set per analyzer is derived from what the downstream authoritative symbol check can possibly match (verified against `src/Common/WellKnown/MoqKnownSymbols.cs`):

   | Analyzer | Authoritative check | Symbol groups | Pre-filter name(s) |
   |---|---|---|---|
   | Moq1206 | `IsMoqReturnsMethod` (`src/Common/ISymbolExtensions.Moq.cs:79-84`) | `IReturnsReturns`, `IReturns1Returns`, `IReturns2Returns` — all created as `CreateLazyMethods(..., "Returns")` (`MoqKnownSymbols.cs:108-110`) | `"Returns"` |
   | Moq1208 | `IsMoqReturnsMethod` (same) | same | `"Returns"` |
   | Moq1201 | `IsMoqSetupMethod` (`ISymbolExtensions.Moq.cs:40-43`) | `Mock1Setup` = `CreateLazyMethods(Mock1, "Setup")` (`MoqKnownSymbols.cs:75`) | `"Setup"` |
   | Moq1202 | `IsInstanceOf(knownSymbols.Mock1Raise)` | `Mock1Raise` = `CreateLazyMethods(Mock1, "Raise")` (`MoqKnownSymbols.cs:82`) | `"Raise"` |
   | Moq1205 | `IsMoqEventSetupMethod` (`ISymbolExtensions.Moq.cs:55-58`) | `Mock1SetupAdd`/`Mock1SetupRemove` = `CreateLazyMethods(Mock1, "SetupAdd"/"SetupRemove")` (`MoqKnownSymbols.cs:79-80`) | `"SetupAdd"`, `"SetupRemove"` |

   **ADR-001 note (state this in the PR):** ADR-001 (`docs/architecture/ADR-001-symbol-based-detection-over-string-matching.md`) forbids name-based *detection*; name-based checks are allowed **only as fast-path pre-filters that run before an authoritative symbol check**. Every pre-filter added here rejects early; nothing is ever *reported* based on a name. This exactly mirrors the existing sanctioned `IsMoqFluentInvocation` fast path.

2. Every `GetSymbolInfo`/`GetTypeInfo` call in the five analyzers and the five Common helper sites passes the analysis `CancellationToken`.

3. Pre-filter shape subtlety (verified against current code — get this right):
   - Moq1206, Moq1208, and Moq1201 **already require** `invocation.Expression is MemberAccessExpressionSyntax` before binding (`ReturnsAsync...:86-89`, `ReturnsDelegate...:86-89`, `SetupShouldNotIncludeAsyncResult...:61-64`), so a `MemberAccessExpressionSyntax { Name.Identifier.ValueText: "..." }` pattern is *exactly* behavior-preserving for them.
   - Moq1202 and Moq1205 currently bind the **whole invocation** with no shape requirement, so `mock?.Raise(...)` (expression is `MemberBindingExpressionSyntax`) is analyzed today. A `MemberAccessExpressionSyntax`-only pre-filter would silently introduce a false negative there. Use a shape-tolerant name extractor that falls through to the symbol check for unknown shapes (code below).

## Implementation plan

All new/changed `CancellationToken` parameters use a trailing optional parameter (`CancellationToken cancellationToken = default`) to match the existing style of `IMethodSymbolExtensions.TryGetParameterOfType` and to avoid touching the direct unit tests of `TryGetEventNameFromLambdaSelector` in `tests/Moq.Analyzers.Test/Common/SemanticModelExtensionsTests.cs:191-260`. Every **production** call site listed below must pass the token explicitly.

### 1. `src/Common/InvocationExpressionSyntaxExtensions.cs`

Add the shared shape-tolerant pre-filter helper and thread the token through `FindSetupInvocation`:

```csharp
    /// <summary>
    /// Fast-path name pre-filter: determines whether an invocation could be calling a member
    /// with the given name, based on syntax alone.
    /// </summary>
    /// <remarks>
    /// Per ADR-001 this is an intentional optimization, not detection logic; callers must
    /// still perform the authoritative symbol check. Unknown expression shapes return
    /// <see langword="true"/> so the symbol check remains the deciding gate.
    /// </remarks>
    /// <param name="invocation">The invocation to check.</param>
    /// <param name="name">The member name to compare against.</param>
    /// <returns><see langword="false"/> only when the invoked member name is known and differs from <paramref name="name"/>.</returns>
    internal static bool CouldInvokeMemberNamed(this InvocationExpressionSyntax invocation, string name)
    {
        return invocation.Expression switch
        {
            MemberAccessExpressionSyntax memberAccess => string.Equals(memberAccess.Name.Identifier.ValueText, name, StringComparison.Ordinal),
            MemberBindingExpressionSyntax memberBinding => string.Equals(memberBinding.Name.Identifier.ValueText, name, StringComparison.Ordinal),
            IdentifierNameSyntax identifier => string.Equals(identifier.Identifier.ValueText, name, StringComparison.Ordinal),
            _ => true, // Unknown shape: defer to the authoritative symbol check.
        };
    }
```

```csharp
    internal static InvocationExpressionSyntax? FindSetupInvocation(this ExpressionSyntax receiver, SemanticModel semanticModel, MoqKnownSymbols knownSymbols, CancellationToken cancellationToken = default)
    {
        ExpressionSyntax current = receiver;

        // Moq fluent chains are short (Setup.Callback.Returns at most 3-4 deep).
        // Guard against pathological syntax trees.
        for (int depth = 0; depth < 10; depth++)
        {
            ExpressionSyntax unwrapped = current.WalkDownParentheses();

            if (unwrapped is not InvocationExpressionSyntax { Expression: MemberAccessExpressionSyntax candidateMemberAccess } candidateInvocation)
            {
                return null;
            }

            SymbolInfo symbolInfo = semanticModel.GetSymbolInfo(candidateInvocation, cancellationToken);
            if (symbolInfo.Symbol != null && symbolInfo.Symbol.IsMoqSetupMethod(knownSymbols))
            {
                return candidateInvocation;
            }

            // Continue walking up the chain (past Callback, etc.)
            current = candidateMemberAccess.Expression;
        }

        return null;
    }
```

Callers to update (pass `context.CancellationToken`): `ReturnsAsyncShouldBeUsedForAsyncMethodsAnalyzer.cs:60`, `ReturnsDelegateShouldReturnTaskAnalyzer.cs:111`.

### 2. `src/Common/SemanticModelExtensions.cs`

Complete replacements (only the token plumbing changes):

```csharp
    internal static IEnumerable<IMethodSymbol> GetAllMatchingMockedMethodSymbolsFromSetupMethodInvocation(this SemanticModel semanticModel, InvocationExpressionSyntax? setupMethodInvocation, CancellationToken cancellationToken = default)
    {
        LambdaExpressionSyntax? setupLambdaArgument = setupMethodInvocation?.ArgumentList.Arguments[0].Expression as LambdaExpressionSyntax;

        return setupLambdaArgument?.Body is not InvocationExpressionSyntax mockedMethodInvocation
            ? []
            : semanticModel.GetAllMatchingSymbols<IMethodSymbol>(mockedMethodInvocation, cancellationToken);
    }

    internal static bool IsCallbackOrReturnInvocation(this SemanticModel semanticModel, InvocationExpressionSyntax callbackOrReturnsInvocation, MoqKnownSymbols knownSymbols, CancellationToken cancellationToken = default)
    {
        return semanticModel.IsMoqFluentInvocation(
            callbackOrReturnsInvocation,
            CallbackOrReturnNames,
            knownSymbols,
            static (symbol, ks) => IsCallbackOrReturnSymbol(symbol, ks),
            cancellationToken);
    }

    internal static bool IsRaisesInvocation(this SemanticModel semanticModel, InvocationExpressionSyntax raisesInvocation, MoqKnownSymbols knownSymbols, CancellationToken cancellationToken = default)
    {
        return semanticModel.IsMoqFluentInvocation(
            raisesInvocation,
            RaisesNames,
            knownSymbols,
            static (symbol, ks) => symbol.IsMoqRaisesMethod(ks),
            cancellationToken);
    }
```

In `GetAllMatchingSymbols<T>` (`:155-193`), change the signature to `private static List<T> GetAllMatchingSymbols<T>(this SemanticModel semanticModel, ExpressionSyntax expression, CancellationToken cancellationToken)` and line `:160` to:

```csharp
        SymbolInfo symbolInfo = semanticModel.GetSymbolInfo(expression, cancellationToken);
```

In `IsMoqFluentInvocation` (`:254-290`), add a trailing `CancellationToken cancellationToken` parameter and change line `:283` to:

```csharp
        SymbolInfo symbolInfo = semanticModel.GetSymbolInfo(memberAccess, cancellationToken);
```

In `TryGetEventNameFromLambdaSelector` (`:105`), `TryGetEventTypeFromLambdaSelector` (`:120`), `TryGetEventPropertyFromLambdaSelector` (`:137`), and `TryGetEventSymbolFromLambdaSelector` (`:202`), add a trailing `CancellationToken cancellationToken = default` (required, non-default on the private ones) and change line `:229` to:

```csharp
        SymbolInfo symbolInfo = semanticModel.GetSymbolInfo(memberAccess, cancellationToken);
```

Do **not** change the `Arguments[0]` indexing at `:60` here — that is a separate audit finding (A-2) with its own issue.

Callers to update (pass `context.CancellationToken`):

- `src/Analyzers/CallbackSignatureShouldMatchMockedMethodAnalyzer.cs:78` (`IsCallbackOrReturnInvocation`) and `:166` (`GetAllMatchingMockedMethodSymbolsFromSetupMethodInvocation`).
- `src/CodeFixes/CallbackSignatureShouldMatchMockedMethodFixer.cs:165` (`GetAllMatchingMockedMethodSymbolsFromSetupMethodInvocation`) — `cancellationToken` is already a parameter of the enclosing `FindSingleMockedMethod` (`:156`).
- `src/Analyzers/RaisesEventArgumentsShouldMatchEventSignatureAnalyzer.cs:69` (`IsRaisesInvocation`).
- `src/Common/EventSyntaxExtensions.cs:150` and `:172` (see step 3).

### 3. `src/Common/EventSyntaxExtensions.cs`

Thread the token through `TryGetEventMethodArguments` (`:87-124`), `TryGetEventMethodArgumentsFromLambdaSelector` (`:136-154`), and `GetEventNameFromSelector` (`:162-175`): add a trailing `CancellationToken cancellationToken = default` to each; change the extractor delegate type to `Func<SemanticModel, ExpressionSyntax, CancellationToken, (bool Success, ITypeSymbol? EventType)>` and invoke it as `eventTypeExtractor(semanticModel, eventSelector, cancellationToken)`; inside `TryGetEventMethodArgumentsFromLambdaSelector`'s static lambda call `sm.TryGetEventTypeFromLambdaSelector(selector, out ITypeSymbol? eventType, ct)`; inside `GetEventNameFromSelector` call `semanticModel.TryGetEventNameFromLambdaSelector(eventSelector, out string? eventName, cancellationToken)`. Update the direct unit tests of the extractor overload in `tests/Moq.Analyzers.Test/Common/EventSyntaxExtensionsTests.cs:244-405` only if the compiler requires it (the delegate type change makes it required — update those test lambdas to take the third parameter).

Callers to update (pass `context.CancellationToken`): `RaiseEventArgumentsShouldMatchEventSignatureAnalyzer.cs:72` and `:77`; `RaisesEventArgumentsShouldMatchEventSignatureAnalyzer.cs:74` and `:79`.

### 4. `src/Common/SyntaxNodeExtensions.cs`

```csharp
    internal static Location? FindLocation<TSyntax>(this SyntaxNode syntax, ISymbol memberSymbol, SemanticModel? semanticModel, CancellationToken cancellationToken = default)
        where TSyntax : SyntaxNode
    {
        TSyntax? node = syntax.DescendantNodes()
            .OfType<TSyntax>()
            .FirstOrDefault(n =>
            {
                ISymbol? symbol = semanticModel?.GetSymbolInfo(n, cancellationToken).Symbol;
                return SymbolEqualityComparer.Default.Equals(symbol, memberSymbol);
            });

        return node != null
            ? Location.Create(syntax.SyntaxTree, node.Span)
            : null;
    }
```

Callers to update: `src/Analyzers/LinqToMocksExpressionShouldBeValidAnalyzer.cs:272` and `:279` — add a `CancellationToken` parameter to the enclosing `GetMemberReferenceLocation` (`:267`) and pass `context.CancellationToken` from its call site at `:206`.

### 5. `src/Analyzers/ReturnsAsyncShouldBeUsedForAsyncMethodsAnalyzer.cs`

Replace `Analyze`, `IsReturnsMethodCallWithAsyncLambda`, and `GetAsyncMethodName`:

```csharp
    private static void Analyze(SyntaxNodeAnalysisContext context, MoqKnownSymbols knownSymbols)
    {
        InvocationExpressionSyntax invocation = (InvocationExpressionSyntax)context.Node;

        // Fast-path name pre-filter (ADR-001: optimization only; the authoritative gate is the
        // IsMoqReturnsMethod symbol check below). All Moq Returns methods are named "Returns",
        // and this analyzer already required a MemberAccessExpressionSyntax, so this cannot
        // change behavior.
        if (invocation.Expression is not MemberAccessExpressionSyntax { Name.Identifier.ValueText: "Returns" } memberAccess)
        {
            return;
        }

        // Check if this is a Returns method call with async lambda
        if (!IsReturnsMethodCallWithAsyncLambda(invocation, context.SemanticModel, knownSymbols, context.CancellationToken))
        {
            return;
        }

        // Find the Setup call that this Returns is chained from
        InvocationExpressionSyntax? setupInvocation = memberAccess.Expression.FindSetupInvocation(context.SemanticModel, knownSymbols, context.CancellationToken);
        if (setupInvocation == null)
        {
            return;
        }

        // Check if the Setup is for an async method and get the method name
        string? methodName = GetAsyncMethodName(setupInvocation, context.SemanticModel, knownSymbols, context.CancellationToken);
        if (methodName == null)
        {
            return;
        }

        // Report diagnostic on just the Returns(...) method call
        // Create a span from the Returns identifier through the end of the invocation
        int startPos = memberAccess.Name.SpanStart;
        int endPos = invocation.Span.End;
        Microsoft.CodeAnalysis.Text.TextSpan span = Microsoft.CodeAnalysis.Text.TextSpan.FromBounds(startPos, endPos);
        Location location = Location.Create(invocation.SyntaxTree, span);

        Diagnostic diagnostic = location.CreateDiagnostic(Rule, methodName);
        context.ReportDiagnostic(diagnostic);
    }

    private static bool IsReturnsMethodCallWithAsyncLambda(InvocationExpressionSyntax invocation, SemanticModel semanticModel, MoqKnownSymbols knownSymbols, CancellationToken cancellationToken)
    {
        // Query the invocation (not the MemberAccessExpressionSyntax) so Roslyn has argument context
        // for overload resolution. Fall back to CandidateSymbols for delegate overloads.
        SymbolInfo symbolInfo = semanticModel.GetSymbolInfo(invocation, cancellationToken);
        bool isReturnsMethod = symbolInfo.Symbol is IMethodSymbol method
            ? method.IsMoqReturnsMethod(knownSymbols)
            : symbolInfo.CandidateSymbols
                .OfType<IMethodSymbol>()
                .Any(m => m.IsMoqReturnsMethod(knownSymbols));

        if (!isReturnsMethod)
        {
            return false;
        }

        // Check if the Returns call has an async lambda argument
        return HasAsyncLambdaArgument(invocation);
    }

    private static string? GetAsyncMethodName(InvocationExpressionSyntax setupInvocation, SemanticModel semanticModel, MoqKnownSymbols knownSymbols, CancellationToken cancellationToken)
    {
        // Check if this is a Setup method call
        if (setupInvocation.Expression is not MemberAccessExpressionSyntax setupMemberAccess)
        {
            return null;
        }

        SymbolInfo setupSymbolInfo = semanticModel.GetSymbolInfo(setupMemberAccess, cancellationToken);
        if (setupSymbolInfo.Symbol is null || !setupSymbolInfo.Symbol.IsMoqSetupMethod(knownSymbols))
        {
            return null;
        }

        // Get the mocked method from the setup
        ExpressionSyntax? mockedMemberExpression = setupInvocation.FindMockedMemberExpressionFromSetupMethod();
        if (mockedMemberExpression == null)
        {
            return null;
        }

        SymbolInfo mockedSymbolInfo = semanticModel.GetSymbolInfo(mockedMemberExpression, cancellationToken);
        if (mockedSymbolInfo.Symbol is not IMethodSymbol mockedMethod)
        {
            return null;
        }

        // Check if the mocked method returns Task or Task<T>
        if (!mockedMethod.ReturnType.IsTaskOrValueTaskType(knownSymbols))
        {
            return null;
        }

        // Return the method name
        return mockedMethod.Name;
    }
```

(The `IsReturnsMethodCallWithAsyncLambda` member-access shape check is removed because the pre-filter in `Analyze` already guarantees it; the local `MemberAccessExpressionSyntax memberAccess = (MemberAccessExpressionSyntax)invocation.Expression;` cast at old `:59` is replaced by the pattern variable.)

### 6. `src/Analyzers/ReturnsDelegateShouldReturnTaskAnalyzer.cs`

- In `Analyze` (`:50`), pass `context.CancellationToken` into `IsReturnsMethodCallWithSyncDelegate` and `TryGetMismatchInfo` (add `CancellationToken cancellationToken` parameters to both, and to `HasSyncDelegateArgument`, `IsMethodGroupExpression`, `GetDelegateReturnType`, `GetReturnTypeFromBlock`).
- In `IsReturnsMethodCallWithSyncDelegate` (`:76-119`), replace the shape check at `:86-89` with the pre-filter:

```csharp
        if (invocation.Expression is not MemberAccessExpressionSyntax { Name.Identifier.ValueText: "Returns" } access)
        {
            return false;
        }
```

  and update the binding calls: `:93` → `semanticModel.GetSymbolInfo(invocation, cancellationToken)`; `:111` → `access.Expression.FindSetupInvocation(semanticModel, knownSymbols, cancellationToken)`.
- `IsMethodGroupExpression` `:149` → `semanticModel.GetSymbolInfo(expression, cancellationToken)`.
- `TryGetMismatchInfo` `:180` → `semanticModel.GetSymbolInfo(mockedMemberExpression, cancellationToken)`.
- `GetDelegateReturnType` `:236` → `semanticModel.GetSymbolInfo(firstArgument, cancellationToken)`.
- `GetReturnTypeFromBlock` `:270` → `semanticModel.GetTypeInfo(returnStatement.Expression, cancellationToken)`.

All other logic stays byte-for-byte identical.

### 7. `src/Analyzers/SetupShouldNotIncludeAsyncResultAnalyzer.cs`

Only the pre-filter (tokens already threaded). Replace the shape check at `:61-64`:

```csharp
        // Fast-path name pre-filter (ADR-001: optimization only; the authoritative gate is the
        // IsMoqSetupMethod symbol check below). Mock<T>.Setup is named "Setup", and this analyzer
        // already required a MemberAccessExpressionSyntax, so this cannot change behavior.
        if (setupInvocation.Expression is not MemberAccessExpressionSyntax { Name.Identifier.ValueText: "Setup" } memberAccessExpression)
        {
            return;
        }
```

### 8. `src/Analyzers/RaiseEventArgumentsShouldMatchEventSignatureAnalyzer.cs`

Replace `Analyze` and `IsRaiseMethodCall`:

```csharp
    private static void Analyze(SyntaxNodeAnalysisContext context, MoqKnownSymbols knownSymbols)
    {
        InvocationExpressionSyntax invocation = (InvocationExpressionSyntax)context.Node;

        // Fast-path name pre-filter (ADR-001: optimization only; the authoritative gate is the
        // Mock1Raise symbol check below). Mock<T>.Raise is named "Raise". This analyzer binds the
        // whole invocation and supports shapes beyond plain member access (e.g. mock?.Raise(...)),
        // so unknown shapes fall through to the symbol check.
        if (!invocation.CouldInvokeMemberNamed("Raise"))
        {
            return;
        }

        // Check if this is a Raise method call on a Mock<T>
        if (!IsRaiseMethodCall(context.SemanticModel, invocation, knownSymbols, context.CancellationToken))
        {
            return;
        }

        if (!EventSyntaxExtensions.TryGetEventMethodArgumentsFromLambdaSelector(invocation, context.SemanticModel, knownSymbols, out ArgumentSyntax[] eventArguments, out ITypeSymbol[] expectedParameterTypes, context.CancellationToken))
        {
            return;
        }

        string eventName = EventSyntaxExtensions.GetEventNameFromSelector(invocation, context.SemanticModel, context.CancellationToken);

        context.ValidateEventArgumentTypes(eventArguments, expectedParameterTypes, invocation, Rule, eventName);
    }

    private static bool IsRaiseMethodCall(SemanticModel semanticModel, InvocationExpressionSyntax invocation, MoqKnownSymbols knownSymbols, CancellationToken cancellationToken)
    {
        SymbolInfo symbolInfo = semanticModel.GetSymbolInfo(invocation, cancellationToken);
        if (symbolInfo.Symbol is not IMethodSymbol methodSymbol)
        {
            return false;
        }

        return methodSymbol.IsInstanceOf(knownSymbols.Mock1Raise);
    }
```

Apply the same two-line change (`IsRaisesInvocation` + `TryGetEventMethodArgumentsFromLambdaSelector`/`GetEventNameFromSelector` token args) in `RaisesEventArgumentsShouldMatchEventSignatureAnalyzer.cs:69-79` — that analyzer already has a name fast path via `IsRaisesInvocation`, so no new pre-filter is needed there.

### 9. `src/Analyzers/EventSetupHandlerShouldMatchEventTypeAnalyzer.cs`

- In `Analyze` (`:48`), add the pre-filter before `IsEventSetupMethodCall`:

```csharp
        // Fast-path name pre-filter (ADR-001: optimization only; the authoritative gate is the
        // IsMoqEventSetupMethod symbol check below).
        if (!invocation.CouldInvokeMemberNamed("SetupAdd") && !invocation.CouldInvokeMemberNamed("SetupRemove"))
        {
            return;
        }
```

- Thread `context.CancellationToken` through: `IsEventSetupMethodCall` (`:68` → `GetSymbolInfo(invocation, cancellationToken)`), `TryGetEventSetupArguments`/`TryGetEventAndHandlerFromSetup` (`:131` → `GetSymbolInfo(memberAccess, cancellationToken)`), `ValidateHandlerType`/`TryGetHandlerTypeFromExpression` (`:175` → `GetSymbolInfo(memberAccess, cancellationToken)`, `:184` → `GetTypeInfo(genericName.TypeArgumentList.Arguments[0], cancellationToken)`, `:191` → `GetTypeInfo(handlerExpression, cancellationToken)`). `ValidateHandlerType` and `TryGetHandlerTypeFromExpression` gain `CancellationToken` parameters (or use `context.CancellationToken` where the `SyntaxNodeAnalysisContext` is already in scope).

## Tests

- **Regression gate:** the entire existing suite must pass unchanged — the pre-filters and token plumbing are behavior-neutral by construction.
- **Doppelganger negative tests** (custom non-Moq class with same-named methods; repo pattern in `tests/Moq.Analyzers.Test/Helpers/DoppelgangerTestHelper.cs`): already exist for four of the five analyzers — `ReturnsAsyncShouldBeUsedForAsyncMethodsAnalyzerTests.cs:145`, `ReturnsDelegateShouldReturnTaskAnalyzerTests.cs:203`, `SetupShouldNotIncludeAsyncResultAnalyzerTests.cs:69`, `EventSetupHandlerShouldMatchEventTypeAnalyzerTests.cs:109` (and `RaisesEventArgumentsShouldMatchEventSignatureAnalyzerTests.cs:243` for the sibling rule). **Missing for `RaiseEventArgumentsShouldMatchEventSignatureAnalyzer` (Moq1202)** — add one:
  1. Add to `DoppelgangerTestHelper.CustomMockClassTemplate` (additive, harmless to other consumers): `public Mock<T> Raise(System.Action<T> eventExpression, params object[] args) => this;`
  2. In `RaiseEventArgumentsShouldMatchEventSignatureAnalyzerTests.cs`, add (mirroring the sibling test at `RaisesEventArgumentsShouldMatchEventSignatureAnalyzerTests.cs:241-249`):

```csharp
    [Theory]
    [MemberData(nameof(DoppelgangerTestHelper.GetAllCustomMockData), MemberType = typeof(DoppelgangerTestHelper))]
    public async Task ShouldPassIfCustomMockClassIsUsed(string mockCode)
    {
        await Verifier.VerifyAnalyzerAsync(
            DoppelgangerTestHelper.CreateTestCode(mockCode),
            ReferenceAssemblyCatalog.Net80WithNewMoq);
    }
```

  and add a data row exercising the custom `Raise`: `["""new Mock<IMyService>().Raise(x => x.Do("test"), "arg1", "arg2");"""]` (no diagnostics expected — the custom `Raise` is not `Moq.Mock<T>.Raise`).
- Use the repo patterns: `{|MoqXXXX:...|}` markup, `[Theory]` + `[MemberData]`, `.WithNamespaces().WithMoqReferenceAssemblyGroups()` where applicable.

## Constraints

- Analyzer assemblies target **netstandard2.0**; Roslyn is pinned to **Microsoft.CodeAnalysis 4.8** per ADR-003 (`docs/architecture/ADR-003-pin-roslyn-sdk-to-codeanalysis-4-8.md`). Use no APIs newer than 4.8. **No new package dependencies.**
- **ADR-001**: name-based checks are permitted **only** as fast-path pre-filters ahead of an authoritative symbol check — which is precisely what this issue adds; the symbol checks remain the deciding gate everywhere.
- **ADR-007** prefers `RegisterOperationAction`; migrating these five analyzers to operation actions is the eventual ADR-preferred fix but is explicitly out of scope here (see below) — the pre-filter is the minimal, low-risk change.
- Tests are mandatory (see above); zero behavior change intended.

## Validation

1. `dotnet format` (no diff).
2. `dotnet build` — must complete with **0 warnings**.
3. `dotnet test --settings ./build/targets/tests/test.runsettings` — all tests pass.
4. Benchmarks exist for four of the five rules (`tests/Moq.Analyzers.Benchmarks/Moq1201AsyncResultBenchmarks.cs`, `Moq1202RaiseEventBenchmarks.cs`, `Moq1206ReturnsAsyncBenchmarks.cs`, `Moq1208ReturnsDelegateBenchmarks.cs`). Run the documented perf flow and compare to baseline (see `build/scripts/perf/README.md` and CONTRIBUTING.md "Performance Testing Guidelines", line 811):

```powershell
./build/scripts/perf/PerfCore.ps1 -projects "tests/Moq.Analyzers.Benchmarks/Moq.Analyzers.Benchmarks.csproj" -diff
```

   CI also runs the `perf` job (required status check) on the PR.

## Acceptance criteria

- [ ] All five analyzers reject non-matching member names before any `GetSymbolInfo` call, with the exact name sets from the table above.
- [ ] Moq1202/Moq1205 pre-filters are shape-tolerant (member access, member binding, identifier; unknown shapes fall through) so `mock?.Raise(...)` behavior is unchanged.
- [ ] Every `GetSymbolInfo`/`GetTypeInfo` call in the five analyzers and in `SemanticModelExtensions.cs` (`:160`, `:229`, `:283`), `InvocationExpressionSyntaxExtensions.cs` (`:42`), and `SyntaxNodeExtensions.cs` (`:22`) receives a `CancellationToken`, and every production caller passes `context.CancellationToken`.
- [ ] Pre-filter comments cite ADR-001 and state the symbol check is authoritative.
- [ ] New Moq1202 doppelganger negative test added; all existing tests pass unchanged.
- [ ] Build has 0 warnings; perf comparison shows no regression (improvement expected on the four benchmarked rules).

## Out of scope

- Migrating these analyzers to `RegisterOperationAction` (ADR-007) — larger refactor, tracked separately.
- The unguarded `Arguments[0]` in `SemanticModelExtensions.cs:60` (audit A-2) — separate issue.
- Deferring the eager `GetEventNameFromSelector`/`ToDisplayString` allocations in the Raise/Raises analyzers (audit A-13/L7) — separate issue ("Defer report-path allocations"); if that issue lands first, resolve trivial merge conflicts in the same methods in favor of both changes.
- `EventSetupHandlerShouldMatchEventTypeAnalyzer.cs:139`'s eager `eventSymbol.ToDisplayString()` — same separate issue territory.
